alephium/caip10.md

```diff
@@ -0,0 +1,74 @@
+---
+namespace-identifier: alephium-caip10
+title: Alephium Namespace - Addresses
+author: Hongchao Liu (@h0ngcha0)
+discussions-to: https://github.com/ChainAgnostic/namespaces/pull/117
+status: Draft
+type: Standard
+created: 2024-06-19
+updated: 2024-06-19
+requires: ["CAIP-2", "CAIP-10"]
+---
+
+# CAIP-10
+
+*For context, see the [CAIP-10][] specification.*
+
+## Rationale
+
+An address on Alephium is a unique identifier that represents an
+account or a contract. All networks share the same address format.
+
+There are 4 different address types on Alephium:
+
+* 0x00 - Pay to public key hash (P2PKH)
+* 0x01 - Pay to multiple public key Hash (P2MPKH)
+* 0x02 - Pay to script hash (P2SH)
+* 0x03 - Pay to contract (P2C)
+
+For each address type, `content bytes` are defined as follows:
+
+- **P2PKH** - serialized public key hash
+- **P2MPKH** - serialized public key hashes and multisig threshold
+- **P2SH** - serialized script hash
+- **P2C** - serialized contract id
+
+Hash algorithm used is blake2b.
+
+Constructing an address:
+
+- **address** = `address type || content bytes`
+
+Each address on Alephium belongs to a group, which can be derived
+deterministically from the address.
+
+## Syntax
+
+Alephium address is represented as a Base58 encoded string.
```

Comment:
```suggestion
The native Alephium address encoding is a [Base58][]-encoded string of the type defined above.
[CAIP-10][] simply uses an entire native address as the `account_address` component.
```

Reply:
But this begs the question... what is the [chain_id segment](https://chainagnostic.org/CAIPs/caip-10#syntax) for each CAIP-10 example listed below in the test vectors?

You mention above that group membership can be derived from the address, but it was unclear how.  Do the address types correspond to group 0, 1, 2, and 3 respectively?

I was also unclear if each address "exists" only on a specific network (i.e., on mainnet or devnet), or if an address is equally valid on all 3, even if it only has history on 1 or 0 of them.

The CAIP-10 system is heirarchical and most CAIP-10 implementations have a usecase or two requiring them to know (if not a universal need to know) _which network to query for history_ of a given account/address, which is why there is a CAIP-2 embedded in each CAIP-10.
If this is not the case and addresses are universal, you might want to add a "-1" to the enum of networks you define in CAIP-2, like `offchain` or `universal`. This way, all CAIP-10s could be `alephium:universal/-1:{address}`.
Conversely, if there IS no universal/offchain address namespace in the app development model of alephium, do NOT introduce such a CAIP-2 wildcard but explain how a given address should be namespaced/mapped to CAIP-2s.

Fascinating stuff here, I love how different this is from other CAIP-2 and CAIP-10 profiles!

Reply:
Thanks for the great comment!

> You mention above that group membership can be derived from the address, but it was unclear how. Do the address types correspond to group 0, 1, 2, and 3 respectively?

I added a description in this [commit](https://github.com/ChainAgnostic/namespaces/pull/117/commits/3afba3c2bfd832319e29de8af9950f59c81b16f0), basically the group information is encoded as part of the address and can be derived from it. 

Thanks for the context for `CAIP-2`. The address is indeed "universal" in the sense that it is valid for all networks, and the group information is implicitly embedded in the address. However dApps on Alephium can request an address from a certain group on a certain network. For example, a testnet dApp can request connection on chain id `alephium:testnet/1`, the wallet should connect the dApp with an address from group `1` on testnet. If a mainnet dApp request connection on chain `alephium:mainnet/-1`, then the wallet can connect the dApp with an address from any groups on mainnet.

So I am a bit unsure here as well, I specified chain id in`CAIP-2` as:
```
alephium + ":" + network_id + "/" + group
```
Because I thought dApp can use this format to request address from a certain group (or all groups) on a certain network, but since address is universal and group is already implicitly embedded in address, if feels like the address should be specified like `alephium:universal:{address}`. 

Really appreciate it if you could give us some advice here :pray:

Reply:
Actually, I realized that `/` won't work in a CAIP-2 reference, the only diacritics allowed are `-` and `_`.  Since `-1` makes sense as the wildcard, that just leaves `_` as the discriminant.

Wouldn't it make more sense, given all this, to use this as the syntax?

`alephium + ":" + group { -1 | 0 | 1 | 2 | 3} + "_" + network_id `

Reply:
Thanks! I have updated the chain id format in this [commit](https://github.com/ChainAgnostic/namespaces/pull/117/commits/b16b4fe480accc5bde28f5134e70f9aabf0e9bb0)